tests/functional/oauth_webchannel_keys.js

```diff
@@ -0,0 +1,510 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define([
+  'intern',
+  'intern!object',
+  'intern/chai!assert',
+  'require',
+  'intern/node_modules/dojo/node!xmlhttprequest',
+  'app/bower_components/fxa-js-client/fxa-client',
+  'tests/lib/restmail',
+  'tests/lib/helpers',
+  'tests/functional/lib/helpers',
+  'tests/functional/lib/fx-desktop'
+], function (intern, registerSuite, assert, require, nodeXMLHttpRequest,
+        FxaClient, restmail, TestHelpers, FunctionalHelpers, FxDesktopHelpers) {
+  'use strict';
+
+  var config = intern.config;
+  var AUTH_SERVER_ROOT = config.fxaAuthRoot;
+  var SYNC_URL = config.fxaContentRoot + 'signin?context=fx_desktop_v1&service=sync';
+
+  var PASSWORD = 'password';
+  var TOO_YOUNG_YEAR = new Date().getFullYear() - 13;
+  var OLD_ENOUGH_YEAR = TOO_YOUNG_YEAR - 1;
+  var user;
+  var email;
+  var client;
+  var ANIMATION_DELAY_MS = 1000;
+
+  var listenForSyncCommands = FxDesktopHelpers.listenForFxaCommands;
+  var testIsBrowserNotifiedOfSyncLogin = FxDesktopHelpers.testIsBrowserNotifiedOfLogin;
+
+
+  /**
+   * This suite tests the WebChannel functionality for delivering encryption keys
+   * in the OAuth signin and signup cases. It uses a CustomEvent "WebChannelMessageToChrome"
+   * to finish OAuth flows
+   */
+
+  function testIsBrowserNotifiedOfLogin(context, options) {
+    options = options || {};
+    return FunctionalHelpers.testIsBrowserNotified(context, 'oauth_complete', function (data) {
+      assert.ok(data.redirect);
+      assert.ok(data.code);
+      assert.ok(data.state);
+      // All of these flows should produce encryption keys.
+      assert.ok(data.keys);
+      assert.equal(data.closeWindow, options.shouldCloseTab);
+    });
+  }
+
+  function openFxaFromRp(context, page, additionalQueryParams) {
+    var queryParams = '&webChannelId=test&keys=true';
+    for (var key in additionalQueryParams) {
+      queryParams += ('&' + key + '=' + additionalQueryParams[key]);
+    }
+    return FunctionalHelpers.openFxaFromRp(context, page, queryParams);
+  }
+
+
+  registerSuite({
+    name: 'oauth web channel keys',
+
+    beforeEach: function () {
+      email = TestHelpers.createEmail();
+      user = TestHelpers.emailToUser(email);
+
+      client = new FxaClient(AUTH_SERVER_ROOT, {
+        xhr: nodeXMLHttpRequest.XMLHttpRequest
+      });
+
+      return FunctionalHelpers.clearBrowserState(this, {
+        contentServer: true,
+        '123done': true
+      });
+    },
+
+    'signup, verify same browser with original tab open': function () {
```

Comment:
Nit: `signup, verify same browser with original tab open'` ->
signup, verify same browser, in a different tab, with original tab open' ?